Refactor absorption_cross_section for NIR Corrections & Update MCfit to Include PAH Weights

STIK98 opened the pull request:
# Refactor `absorption_cross_section` for NIR Corrections & Update `MCfit` to Include PAH Weights
Hi! My name is Florence Stikkelbroeck, and I’m a second-year Master’s student at the University of Amsterdam. I’ve been working with Dr. Alessandra Candian on my Master’s thesis, where I aim to better constrain the PAH population in Titan’s upper atmosphere. As part of my research, I’ve been using the NASA Ames PAHdb and its accompanying Python Suite extensively over the past few months.

While working with the database, I noticed a couple of areas where small improvements could enhance accuracy and usability. Specifically, the UV absorption cross-section calculation did not include Mattioda’s NIR correction from 2008, which affects the longer-wavelength range, and it referenced the wrong Li & Draine paper. Additionally, I added the ability to extract PAH weights from the Monte Carlo fit, making it possible to calculate a concentration-weighted mean. Two additions which I believe could be useful for further analysis.

These changes will refine the tool and ensure it aligns with the latest available data. Below is a summary of the modifications.

## Summary  
This PR introduces two main updates:  

1. **Refactored `absorption_cross_section` in `transitions.py`** to incorporate Near-Infrared (NIR) corrections from **Mattioda et al. (2008)**. The original function used Mattioda et al. (2005c) which was updated in the 2008 paper, leading to discrepancies at longer wavelengths. This update ensures a more accurate representation of the PAH absorption cross-section by combining:  
   - **UV-VIS Absorption:** Li & Draine (2001), Eq. 11, with values from (2007) table 1.
   - **NIR Correction:** Mattioda et al. (2008), using a single Gaussian approximation  
   - The updated function aligns with the NASA Ames PAHdb Python Suite, which uses **Li & Draine (2001)** for UV-VIS.  

2. **Updated `MCfit` in `mcfitted.py` to return PAH UIDs and weights**, making it easier to analyze the contribution of individual PAHs in spectral fitting.  

## Related Issues  
- Fixes overestimation of absorption cross-section at longer wavelengths.  
- Improves transparency in `MCfit` output.  

## Changes  
- **`absorption_cross_section` updates:**  
  - Removed old Gaussian parameters (Mattioda 2005c).  
  - Implemented **Drude summation for the first two elements** (Li&Draine 2001/2007) and **one Gaussian for NIR correction** (Mattioda 2008).  
  - Added wavelength cutoff correction for **nc > 40** (Mattioda 2005c).  
- **`MCfit` refactor in `mcfitted.py`:**  
  - Now returns a structured dictionary with **PAH UIDs and weights**.  

## Results  

### Absorption Cross-Section Comparison  

The updated function includes NIR corrections, leading to an absorption drop-off in line with the latest research.  

**Before vs. After:**  
![ACS_comparisons](https://github.com/user-attachments/assets/fcf596a7-983f-421e-b7ed-f4d9bc8d49f8)
![chargedvsneutral(1)](https://github.com/user-attachments/assets/955a58d3-9736-4a61-a8a8-8895c2c3377a)

- The update follows **Mattioda et al. (2008)** while maintaining consistency with **Li & Draine (2001, 2007)**.  

## References  
- **Li & Draine (2001)**, ApJ, 554:778-802, Eq. [11]  
- **Draine & Li (2007)**, ApJ, 657:810-837, Table 1  
- **Mattioda et al. (2005c)**  Apj, 629(2):1183 
- **Mattioda et al. (2008)**, ApJ,  680(2):1243 

Let me know if you have any questions or feedback, I’d love to hear your thoughts!
Looking forward to your review.

Kind regards,
Florence

STIK98 committed:
Cross_section update

Made the Crosssection calculation similar to that of López-Puertas et al. 2013

STIK98 committed:
Turned off some of the messages

For readability, I commented out or deleted some of the standard messages that would appear every time I ran the code.

STIK98 committed:
add use_nnls = True

Added Boolian to make NNLS fitting happen even with obs.uncertainty present, in line with Lopez-Puertas (2013)

STIK98 committed:
Update transitions.py

fixed type in 'blackbody'

STIK98 committed:
Update transitions.py

fixed another typo in 'temperature'

STIK98 committed:
Create Testing_ACS.ipynb

STIK98 committed:
Adjusted Absorption_Cross_section

adjusted the original code from the nasa ames pahdb database and fixed two small errors such as correct references to Draine & Li 2007 and Li & Draine 2001. implemented the nc>40 requirement for the cutoff frequency, and replaced the charge =!0 with the updated 2008 code.

STIK98 committed:
Refactor get method to return detailed mcfits structure with PAH UIDs and weights

STIK98 committed:
Remove option to use NNLS with uncertainties in Spectrum class, original file

STIK98 committed:
Revert "Turned off some of the messages"

PAHdb commented:
Hi Florence!

Thank you for contributing to the _AmesPAHdbPythonsSuite_. Your contributions are valuable ones. Unfortunately, due to an outdated package used by the CI the checks failed. Can you please rebase your branch to the latest so they can run. Also, to allow reproducibility of any previous results, it would be good to have the outdated cross sections still available. Perhaps rename the earlier version of `absorption_cross_section` to `absorption_cross_section_MAH05` and have a keyword, e.g., `crosssections="MAH05"`, allow selecting it in the models. Thoughts? Lastly, please consider fixing any failing tests and adding any applicable new ones.

Thank you again!
Christiaan

PAHdb commented on amespahdbpythonsuite/mcfitted.py:
```diff
@@ -52,16 +52,22 @@ def __set(self, d: Optional[dict] = None, **keywords) -> None:
-        return d
+        return {
+            "type": self.__class__.__name__,
+            "mcfits": [
```
Originally "mcfits" held a list of Fiitted objects, allowing one to access all information wrt an individual fit. This functionality is now lost. I addition, one could obtain the uids and weights from each individual fit using the following code:

```python
fits = mcfit.get()
uids = [fit.getuids() for fit in fits["mcfits"]]
weights =[fit.getweights() for fit in fits["mcfits"]]
```

PAHdb commented on amespahdbpythonsuite/transitions.py:
```diff
@@ -920,8 +920,13 @@ def plot(self, **keywords) -> None:
     @staticmethod
     def absorption_cross_section(f: np.ndarray) -> np.ndarray:
```
It would be could to still have the old functionality available as an option--possibly as a separate function and allowing selecting it using a keyword in the emission models.

STIK98 commented:
Thank you for your comment! This was an oversight on my part, but you are absolutely correct. I will attempt to implement this functionality again in the code.

From: NASA Ames PAH IR Spectroscopic Database ***@***.***>
Date: Saturday, 15 February 2025 at 20:20
To: PAHdb/AmesPAHdbPythonSuite ***@***.***>
Cc: Floor Stikkelbroeck ***@***.***>, Author ***@***.***>
Subject: Re: [PAHdb/AmesPAHdbPythonSuite] Refactor absorption_cross_section for NIR Corrections & Update MCfit to Include PAH Weights (PR #23)


@PAHdb commented on this pull request.

________________________________

In amespahdbpythonsuite/mcfitted.py<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#discussion_r1957176610>:

>
-        return d
+        return {
+            "type": self.__class__.__name__,
+            "mcfits": [


Originally "mcfits" held a list of Fiitted objects, allowing one to access all information wrt an individual fit. This functionality is now lost. I addition, one could obtain the uids and weights from each individual fit using the following code:

fits = mcfit.get()
uids = [fit.getuids() for fit in fits["mcfits"]]
weights =[fit.getweights() for fit in fits["mcfits"]]

________________________________

In amespahdbpythonsuite/transitions.py<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#discussion_r1957177370>:

> @@ -920,8 +920,13 @@ def plot(self, **keywords) -> None:
     @staticmethod
     def absorption_cross_section(f: np.ndarray) -> np.ndarray:


It would be could to still have the old functionality available as an option--possibly as a separate function and allowing selecting it using a keyword in the emission models.

—
Reply to this email directly, view it on GitHub<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#pullrequestreview-2619481225>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASMXP6WAWWQM34NUYNJM3D32P6HPDAVCNFSM6AAAAABXCM43SCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMJZGQ4DCMRSGU>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

STIK98 commented:
Dear Christiaan,

Thank you for your detailed feedback and for guiding me through the process. As this is my first time working with GitHub in its full capacity, I appreciate your patience!

I will rebase my branch to the latest version to resolve the CI issues and ensure that the tests run properly. Keeping the previous cross-section calculation for reproducibility makes a lot of sense; I will rename it to “absorption_cross_section_MAH05" and add a keyword argument (e.g., crosssections="MAH05") to allow selection in the models.

I’ll also go through the failing tests and add any necessary new ones to validate the changes. I’ll update you once these adjustments are in place. Thanks again for your help!

Best,
Florence

From: NASA Ames PAH IR Spectroscopic Database ***@***.***>
Date: Saturday, 15 February 2025 at 18:02
To: PAHdb/AmesPAHdbPythonSuite ***@***.***>
Cc: Floor Stikkelbroeck ***@***.***>, Author ***@***.***>
Subject: Re: [PAHdb/AmesPAHdbPythonSuite] Refactor absorption_cross_section for NIR Corrections & Update MCfit to Include PAH Weights (PR #23)


Hi Florence!

Thank you for contributing to the AmesPAHdbPythonsSuite. Your contributions are valuable ones. Unfortunately, due to an outdated package used by the CI the checks failed. Can you please rebase your branch to the latest so they can run. Also, to allow reproducibility of any previous results, it would be good to have the outdated cross sections still available. Perhaps rename the earlier version of absorption_cross_section to absorption_cross_section_MAH05 and have a keyword, e.g., crosssections="MAH05", allow selecting it in the models. Thoughts? Lastly, please consider fixing any failing tests and adding any applicable new ones.

Thank you again!
Christiaan

—
Reply to this email directly, view it on GitHub<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#issuecomment-2661005002>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASMXP6QWAGKDA46PLO5IWXD2P5XJ3AVCNFSM6AAAAABXCM43SCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGAYDKMBQGI>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

[PAHdb]PAHdb left a comment (PAHdb/AmesPAHdbPythonSuite#23)<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#issuecomment-2661005002>

Hi Florence!

Thank you for contributing to the AmesPAHdbPythonsSuite. Your contributions are valuable ones. Unfortunately, due to an outdated package used by the CI the checks failed. Can you please rebase your branch to the latest so they can run. Also, to allow reproducibility of any previous results, it would be good to have the outdated cross sections still available. Perhaps rename the earlier version of absorption_cross_section to absorption_cross_section_MAH05 and have a keyword, e.g., crosssections="MAH05", allow selecting it in the models. Thoughts? Lastly, please consider fixing any failing tests and adding any applicable new ones.

Thank you again!
Christiaan

—
Reply to this email directly, view it on GitHub<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#issuecomment-2661005002>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASMXP6QWAGKDA46PLO5IWXD2P5XJ3AVCNFSM6AAAAABXCM43SCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGAYDKMBQGI>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

STIK98 committed:
Merge branch 'PAHdb:master' into master

PAHdb commented:
Dear Florence,

As you will have seen, some of the checks failed. While many are pedantic, they do help a consistent coding style and are able to catch errors early on. Waiting for the checks to finish on GitHub and see whether they passed can be an involved and frustrating process. To that end, I would recommend running the workflow commands locally prior committing changes. You can find the commands listed in the [YAML](https://github.com/PAHdb/AmesPAHdbPythonSuite/blob/5352daea2b72e4d91f840b770f09076b70c8e7e8/.github/workflows/ci.yml#L1) file found in the `.github/workflows`-directory. For example, to check for `flake8`-errors (which you currently have) you can run:

```sh
flake8 --max-line-length=115 --exclude "versioneer.py,docs/*"
``` 

Take care and best wishes,
Christiaan

STIK98 commented:
Dear Christiaan,

Thank you for the suggestion! I hadn’t thought of running the checks locally first, but that makes a lot of sense. I’ll review the YAML file to see the full workflow and make sure everything passes before committing again. I think my version is up to date now with the most recent commit from your end.

I’ll also run `flake8` and fix the style issues. Thanks again for your help!

Best,
Florence


From: NASA Ames PAH IR Spectroscopic Database ***@***.***>
Date: Wednesday, 19 February 2025 at 18:47
To: PAHdb/AmesPAHdbPythonSuite ***@***.***>
Cc: Floor Stikkelbroeck ***@***.***>, Author ***@***.***>
Subject: Re: [PAHdb/AmesPAHdbPythonSuite] Refactor absorption_cross_section for NIR Corrections & Update MCfit to Include PAH Weights (PR #23)


Dear Florence,

As you will have seen, some of the checks failed. While many are pedantic, they do help a consistent coding style and are able to catch errors early on. Waiting for the checks to finish on GitHub and see whether they passed can be an involved and frustrating process. To that end, I would recommend running the workflow commands locally prior committing changes. You can find the commands listed in the YAML<https://github.com/PAHdb/AmesPAHdbPythonSuite/blob/5352daea2b72e4d91f840b770f09076b70c8e7e8/.github/workflows/ci.yml#L1> file found in the .github/workflows-directory. For example, to check for flake8-errors (which you currently have) you can run:

flake8 --max-line-length=115 --exclude "versioneer.py,docs/*"

Take care and best wishes,
Christiaan

—
Reply to this email directly, view it on GitHub<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#issuecomment-2669350547>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASMXP6UZ4BM2GKKWT2UIQPD2QS7SNAVCNFSM6AAAAABXCM43SCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRZGM2TANJUG4>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

[PAHdb]PAHdb left a comment (PAHdb/AmesPAHdbPythonSuite#23)<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#issuecomment-2669350547>

Dear Florence,

As you will have seen, some of the checks failed. While many are pedantic, they do help a consistent coding style and are able to catch errors early on. Waiting for the checks to finish on GitHub and see whether they passed can be an involved and frustrating process. To that end, I would recommend running the workflow commands locally prior committing changes. You can find the commands listed in the YAML<https://github.com/PAHdb/AmesPAHdbPythonSuite/blob/5352daea2b72e4d91f840b770f09076b70c8e7e8/.github/workflows/ci.yml#L1> file found in the .github/workflows-directory. For example, to check for flake8-errors (which you currently have) you can run:

flake8 --max-line-length=115 --exclude "versioneer.py,docs/*"

Take care and best wishes,
Christiaan

—
Reply to this email directly, view it on GitHub<https://github.com/PAHdb/AmesPAHdbPythonSuite/pull/23#issuecomment-2669350547>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASMXP6UZ4BM2GKKWT2UIQPD2QS7SNAVCNFSM6AAAAABXCM43SCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRZGM2TANJUG4>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

STIK98 committed:
Working gof with chi2

Commiting working version while now trying to calculate the degrees of freedom.

STIK98 committed:
Added reduced chi2

STIK98 committed:
Merge branch 'feature-tests'
